domain/application/types.go

```diff
@@ -293,3 +295,165 @@ type ApplicationConfig struct {
 type ApplicationSettings struct {
 	Trust bool
 }
+
+// Constraints represents the application constraints.
+// All fields of this struct are taken from core/constraints except for the
+// spaces, which are modeled with SpaceConstraint to take into account the
+// negative constraint (excluded field in the db).
+type Constraints struct {
```

Comment:
All these new types and the respective tests were taken from https://github.com/juju/juju/pull/18820.

Reply:
I think it is a mistake to have two copies of the same type and converters in different domains here. 

Even having one is questionable for the lack of lifting we get out of it, but I can set that aside for now.

Having the copies is incorrect modeling, because they are _not_ different domain concerns. At the end of the day they are always applied as _machine_ constraints.

@tlm @SimonRichardson we need to have a think about this.

Reply:
Just move to `domain/constraints`. There is precedence for this (life is a good example).

Reply:
Yes, I think this is a very similar case to life. I created `domain/constraints` and move them there.